ffs/filcold/filcold.go

```diff
@@ -77,50 +81,80 @@ func (fc *FilCold) Fetch(ctx context.Context, pyCid cid.Cid, piCid *cid.Cid, wad
 	return ffs.FetchInfo{RetrievedMiner: miner, FundsSpent: fundsSpent}, nil
 }
 
+func (fc *FilCold) calculatePieceSize(ctx context.Context, c cid.Cid) (api.DataSize, error) {
+	// Limiting deal size calculation since is very resource intensive.
+	// In the next Lotus version (v1.1.3) a new API will be used which also calculates
+	// CommP, so we can help Lotus avoid recalculating size and CommP when deals are made.
+	select {
+	case fc.semaphDealPrep <- struct{}{}:
```

Comment:
I am sure I'm just missing some understanding about how this works, but I would have expected to see some goroutine spawned off after the `fc.semaphDealPrep` is used here to block execution until a slot opens up. As is, it looks to me like calls to `calculatePieceSize` are run serially. Like I said, I'm sure I just don't understand, but I'd like to understand this because it seems awesome!

Reply:
When the scheduler decides to run a Job, it [spawns](https://github.com/textileio/powergate/blob/4b54223d97c4be65fb089c368b0033760975f064/ffs/scheduler/scheduler.go#L379) a new goroutine and let that goroutine execute all the logic.

This means that at this point, when executing cold storage logic, it should only wait for the "OK" of this semaphore to keep doing the work it was doing.

If you see what's happening in a bigger scope, you might see multiple goroutines (Jobs) waiting in this semaphore to be allowed to move on executing heavy stuff.

The good part of this design is that you shouldn't think about the job execution logic methods as single goroutine doing things for multiple jobs. You can always assume you're doing things for a Job already in its own goroutine.

That's why you see that `calculatePieceSize` is serial. Is serial in a single Job execution, but there're multiple goroutines one per executing Job. Whatever buffer this semaphore allows, are parallel executing goroutine Jobs that will be running in parallel calculating their size.

Makes sense?

Reply:
Yea that totally makes sense. I forgot that gouroutines for jobs are created quite early and high up the stack. Yea this is a really nice design because it makes this code look very simple and easy to understand.

Reply:
Exactly!